### PR TITLE
Implement useBlocState and useBlocListener hooks

### PR DESCRIPTION
**Pull request by KamilSztandur**
image

### CODE REVIEW

**Comment by Albert221** [comment]
Why useBlocState when we have context.watch/select?

**Comment by KamilSztandur** [comment]
> Why useBlocState when we have context.watch/select?

Because it has proven itself useful and convenient across many project I worked on. It looks good inside HookWidgets body in terms of convention (stateless -> read/watch/select, hook -> useBloc hooks) especially for locally created cubits (`useBloc` + `useBlocState` + `useBlocListener`) and solves problem of context.watch not refreshing it properly when provided with `Provider` instead of `BlocProvider`

**Review by bartekpacia — COMMENTED** [review]
Let's write a few tests.

File: packages/leancode_hooks/lib/src/use_bloc_state.dart [hunk 1]
```diff
@@ -0,0 +1,12 @@
+import 'package:bloc/bloc.dart';
+import 'package:flutter_hooks/flutter_hooks.dart';
+
+/// Provides a [Cubit] or a [Bloc] current state. Forces [HookWidget] rebuild on
+/// state changed.
```
Comment (shilangyu) [inline packages/leancode_hooks/lib/src/use_bloc_state.dart:5]:
```suggestion
/// Provides [Cubit]'s or [Bloc]'s current state. Forces [HookWidget] to
/// rebuild on state changed.
```

File: packages/leancode_hooks/lib/src/use_bloc_listener.dart [hunk 2]
```diff
@@ -0,0 +1,17 @@
+import 'package:bloc/bloc.dart';
+import 'package:flutter_hooks/flutter_hooks.dart';
+
+/// Takes a [Bloc] or [Cubit] and invokes [listener] in response to state
+/// changes in it.
```
Comment (shilangyu) [inline packages/leancode_hooks/lib/src/use_bloc_listener.dart:5]:
```suggestion
/// Takes a [Bloc] or [Cubit] and invokes [listener] in response to state
/// changes.
```

**Comment by shilangyu** [comment]
I agree with what you said Kamil (that it is useful for local cubits that are not provided), but the last argument bad

> and solves problem of context.watch not refreshing it properly when provided with Provider instead of BlocProvider

Thats not a solution to this problem. The solution is to use `BlocProvider`. If you are not getting refreshes by using `context.watch` you should fix your provider, not switch to useBlocState.

**Review by shilangyu — COMMENTED** [review]
Changelog/pubspec version bump missing

File: packages/leancode_hooks/lib/src/use_bloc_state.dart (hunk 1 above)
Comment (bartekpacia) [reply packages/leancode_hooks/lib/src/use_bloc_state.dart:5]: what about "on state changed" -> "on state changes"/"on state change"?

Comment (shilangyu) [reply packages/leancode_hooks/lib/src/use_bloc_state.dart:5]: > on state changed

Right, missed that. Edited my suggestion

**Comment by shilangyu** [comment]
What is the status of this PR?

**Comment by shilangyu** [comment]
pls lets not leave stale PRs, bump @KamilSztandur

**Comment by KamilSztandur** [comment]
I will wrap it up this week

File: packages/leancode_hooks/lib/src/use_bloc_listener.dart [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+  required void Function(S state) listener,
+}) {
+  useEffect(
+    () {
+      final subscription = bloc.stream.listen(listener);
+      return subscription.cancel;
+    },
+    [bloc, listener],
+  );
+}
```
Comment (shilangyu) [inline packages/leancode_hooks/lib/src/use_bloc_listener.dart:17]: Flutter hooks v0.20.0 has a built in stream listener: `useOnStreamChange`

File: packages/leancode_hooks/lib/src/use_bloc.dart [hunk 4]
```diff
@@ -1,8 +1,8 @@
 import 'package:bloc/bloc.dart';
 import 'package:flutter_hooks/flutter_hooks.dart';
 
-/// Provides a [Cubit] or a [Bloc] that is automatically disposed without having
-/// to use BlocProvider.
+/// Provides [Cubit]'s or [Bloc]'s current state. Forces [HookWidget] to
+/// rebuild on state change.
```
Comment (shilangyu) [inline packages/leancode_hooks/lib/src/use_bloc.dart:5]: I don't think you meant to edit that

File: packages/leancode_hooks/CHANGELOG.md [hunk 5]
```diff
@@ -1,3 +1,8 @@
+# 0.0.5
+
+- Implement `useBlocListener` hook.
+- Implement `useBloc` hook.
```
Comment (shilangyu) [inline packages/leancode_hooks/CHANGELOG.md:4]:
```suggestion
- Add `useBlocListener` hook.
- Add `useBloc` hook.
```

File: packages/leancode_hooks/lib/src/use_bloc_listener.dart (hunk 3 above)
Comment (KamilSztandur) [reply packages/leancode_hooks/lib/src/use_bloc_listener.dart:17]: But we use 0.18.3

**Review by shilangyu — APPROVED** [review]

**Pull request merged**